File: packages/wcss-cli/src/commands/format.ts
import * as fs from 'fs';
import { glob } from 'glob';
import { format } from '../compiler';
import { displaySuccess, displayInfo, displayErrors } from '../error-display';

export interface FormatOptions {
  input: string;
  write: boolean;
}

/**
 * Format command: format WCSS files
 */
export async function formatCommand(options: FormatOptions): Promise<void> {
  const { input, write } = options;

  try {
    // Resolve input files
    const files = await resolveInputFiles(input);

    if (files.length === 0) {
      console.error(`No WCSS files found matching: ${input}`);
      process.exit(1);
    }

    displayInfo(`Formatting ${files.length} file${files.length > 1 ? 's' : ''}...`);

    let errorCount = 0;

    // Format each file
    for (const file of files) {
      try {
        await formatFile(file, write);
      } catch (error) {
        console.error(`Error formatting ${file}:`, error);
        errorCount++;
      }
    }

    // Display summary
    if (errorCount > 0) {
      console.error(`\nFormatting failed for ${errorCount} file${errorCount > 1 ? 's' : ''}`);
      process.exit(1);
    }

    if (write) {
      displaySuccess(`Formatted ${files.length} file${files.length > 1 ? 's' : ''}`);
    } else {
      displayInfo('Run with --write to save changes');
    }
  } catch (error) {
    console.error('Format failed:', error);
    process.exit(1);
  }
}

/**
 * Resolve input files from glob pattern
 */
async function resolveInputFiles(input: string): Promise<string[]> {
  // If input is a directory, search for .wcss files
  if (fs.existsSync(input) && fs.statSync(input).isDirectory()) {
    input = `${input}/**/*.wcss`;
  }

  // Resolve glob pattern
  const files = await glob(input, {
    ignore: ['**/node_modules/**', '**/dist/**'],
  });

  return files;
}

/**
 * Format a single file
 */
async function formatFile(filePath: string, write: boolean): Promise<void> {
  // Read file
  const source = fs.readFileSync(filePath, 'utf-8');

  // Format
  const formatted = await format(source);

  if (write) {
    // Write formatted output
    fs.writeFileSync(filePath, formatted, 'utf-8');
    displayInfo(`Formatted: ${filePath}`);
  } else {
    // Display formatted output
    console.log(`\n${filePath}:`);
    console.log(formatted);
  }
}

File: crates/euis-cli/src/commands/format.rs
use std::fs;

use colored::Colorize;

use super::build::resolve_input_files;

/// Run the format command.
pub fn run(input: &str, write: bool) -> Result<(), String> {
    let files = resolve_input_files(input)?;

    if files.is_empty() {
        return Err(format!("No Euis files found matching: {input}"));
    }

    println!(
        "{} Formatting {} file{}...",
        "info:".cyan(),
        files.len(),
        if files.len() > 1 { "s" } else { "" }
    );

    let mut error_count = 0usize;

    for file in &files {
        if let Err(e) = format_file(file, write) {
            eprintln!("{} Error formatting {}: {}", "error:".red(), file, e);
            error_count += 1;
        }
    }

    if error_count > 0 {
        return Err(format!(
            "Formatting failed for {error_count} file{}",
            if error_count > 1 { "s" } else { "" }
        ));
    }

    if write {
        println!(
            "{} Formatted {} file{}",
            "success:".green(),
            files.len(),
            if files.len() > 1 { "s" } else { "" }
        );
    } else {
        println!("{} Run with --write to save changes", "info:".cyan());
    }

    Ok(())
}

/// Format a single Euis file.
fn format_file(file_path: &str, write: bool) -> Result<(), String> {
    let source =
        fs::read_to_string(file_path).map_err(|e| format!("Failed to read {file_path}: {e}"))?;

    let formatted = euis_compiler::format(&source).map_err(|errors| {
        let msgs: Vec<String> = errors.iter().map(|e| e.message.clone()).collect();
        msgs.join(", ")
    })?;

    if write {
        fs::write(file_path, &formatted)
            .map_err(|e| format!("Failed to write {file_path}: {e}"))?;
        println!("  {} Formatted: {}", "info:".cyan(), file_path);
    } else {
        println!("\n{}:", file_path);
        println!("{}", formatted);
    }

    Ok(())
}
